File: joiner-core/src/main/java/cz/encircled/joiner/Q.java
package cz.encircled.joiner;

import java.util.ArrayList;
import java.util.List;

import com.mysema.query.types.EntityPath;
import com.mysema.query.types.Predicate;

/**
 * @author Kisel on 11.01.2016.
 */
public class Q<T> {

    private Predicate predicate;

    private EntityPath<T> rootEntityPath;

    private List<JoinDescription> joins = new ArrayList<JoinDescription>();

    public static <T> Q<T> from(EntityPath<T> from) {
        return new Q<T>().rootEntityPath(from);
    }

    public Predicate getPredicate() {
        return predicate;
    }

    public Q<T> predicate(Predicate predicate) {
        this.predicate = predicate;
        return this;
    }

    public EntityPath<T> getRootEntityPath() {
        return rootEntityPath;
    }

    public Q<T> rootEntityPath(EntityPath<T> rootEntityPath) {
        this.rootEntityPath = rootEntityPath;
        return this;
    }

    public List<JoinDescription> getJoins() {
        return joins;
    }

    public Q<T> joins(List<JoinDescription> joins) {
        this.joins = joins;
        return this;
    }

}

File: joiner-core/src/main/java/cz/encircled/joiner/query/Q.kt
package cz.encircled.joiner.query

import com.mysema.query.types.EntityPath
import com.mysema.query.types.Expression
import com.mysema.query.types.Predicate
import cz.encircled.joiner.query.join.JoinDescription
import cz.encircled.joiner.util.Assert

import java.util.*

/**
 * TODO create JoinerQuery class
 * This class is a transfer object for repository queries.
 *
 *
 * T - root entity type
 *

 * @author Kisel on 11.01.2016.
 * *
 * @see JoinDescription
 */
class Q<T> internal constructor(val from: EntityPath<T>) {

    var where: Predicate? = null
        private set

    val joins = LinkedHashSet<JoinDescription>()
        get

    val joinGraphs: MutableCollection<String> = ArrayList()

    var isDistinct = true
        private set

    var groupBy: Expression<*>? = null
        private set

    var having: Predicate? = null
        private set

    val hints = LinkedHashMap<String, MutableCollection<Any>>(2)

    val features: MutableCollection<QueryFeature> = ArrayList(2)

    fun distinct(isDistinct: Boolean): Q<T> {
        this.isDistinct = isDistinct
        return this
    }

    fun groupBy(groupBy: Expression<*>): Q<T> {
        this.groupBy = groupBy
        return this
    }

    fun where(where: Predicate): Q<T> {
        this.where = where
        return this
    }

    fun having(having: Predicate): Q<T> {
        this.having = having
        return this
    }

    fun joinGraphs(vararg names: String): Q<T> {
        Collections.addAll(joinGraphs, *names)

        return this
    }

    fun joins(vararg joins: JoinDescription): Q<T> {
        return joins(Arrays.asList(*joins))
    }

    fun joins(joins: Collection<JoinDescription>): Q<T> {
        Assert.notNull(joins)

        for (join in joins) {
            if (!this.joins.add(join)) {
                joins(join.children)
            }
        }

        return this
    }

    fun addHint(hint: String, value: Any): Q<T> {
        hints.computeIfAbsent(hint) { h -> ArrayList<Any>(2) }
        hints[hint]!!.add(value)

        return this
    }

    fun addFeatures(vararg features: QueryFeature): Q<T> {
        Collections.addAll(this.features, *features)
        return this
    }

    companion object {

        fun <T> from(from: EntityPath<T>): Q<T> {
            return Q(from)
        }
    }

}
